src/main/ngapp/accio/src/app/img-gallery/img-gallery.component.ts

```diff
@@ -132,10 +134,171 @@ export class ImgGalleryComponent implements OnInit {
+      this.displayImages = true;
+    }
+  }
+
+  /**
+   * Download a single image or mask.
+   */
+  downloadImage(image: any): void {
+    saveAs(image.url, image.name + '.' + image.type);
+  }
+
+  /** 
+   * Get the Base64 representation of image for zipping.
+   */
+  getBase64String(img: any): string {
+    let canvas = document.createElement('canvas');
+    canvas.width = img.width;
+    canvas.height = img.height;
+    let ctx = canvas.getContext('2d');
+    ctx.drawImage(img, 0, 0);
+    let dataURL = canvas.toDataURL('image');
+    return dataURL.replace(
+      /^data:image\/(png|jpg|gif|pjpeg|bmp|x-icon|svg+xml|webp);base64,/,
+      ''
+    );
+  }
+
+  /**
+   * Download a zip including an image and its masks.
+   */
+  async downloadImageAndItsMasks(image: any): Promise<any> {
+    let zip = new JSZip();
+
+    let img = new Image();
+    img.crossOrigin = 'Anonymous';
+    img.src = image.url;
+    let loadImageString: Promise<string> = new Promise((resolve, reject) => {
+      img.onload = () => resolve(this.getBase64String(img));
+    });
+    zip.file(image.name + '.' + image.type, await loadImageString, {
+      base64: true,
+    });
+
+    this.imgName = image.name;
+    this.maskName = '';
+    this.withMasks = true;
+    this.sortImg = '';
+    this.sortMask = '';
+    this.tag = '';
+
+    let images: any;
+    images = await this.fetchImages();
+
+    for (let i = 0; i < images[0]['masks'].length; i++) {
+      let mask = images[0]['masks'][i];
+      let msk = new Image();
+      msk.crossOrigin = 'Anonymous';
+      msk.src = mask.url;
+      let loadMaskString: Promise<string> = new Promise((resolve, reject) => {
+        msk.onload = () => resolve(this.getBase64String(msk));
+      });
+
+      zip.file('masks/' + mask.name + '.' + mask.type, await loadMaskString, {
+        base64: true,
+      });
+    }
+
+    zip.generateAsync({ type: 'blob' }).then(function (content) {
+      saveAs(content, image.name + '.zip');
+    });
+  }
+
+  /**
+   * Download a zip of all images and masks for the project.
+   */
+  async downloadProjectImages(): Promise<any> {
+    let zip = new JSZip();
+
+    this.imgName = '';
+    this.maskName = '';
+    this.withMasks = true;
+    this.sortImg = '';
+    this.sortMask = '';
+    this.tag = '';
+
+    let images: any;
+    images = await this.fetchImages();
+
+    for (let i = 0; i < images.length; i++) {
+      let image = images[i];
+      let img = new Image();
+      img.crossOrigin = 'Anonymous';
+      img.src = image.url;
+      let loadImageString: Promise<string> = new Promise((resolve, reject) => {
+        img.onload = () => resolve(this.getBase64String(img));
+      });
+      zip.file(
+        image.name + '/' + image.name + '.' + image.type,
+        await loadImageString,
+        {
+          base64: true,
+        }
+      );
+      for (let j = 0; j < images[i]['masks'].length; j++) {
+        let mask = images[i]['masks'][j];
+        let msk = new Image();
```

Comment:
You shouldn't have variables "mask" and "msk"... that's just confusing. Maybe this one can be called "maskImage"?